### ISSUE TITLE
Support user-assigned managed identity in AzureAppConfigurationBuilder

### ISSUE THREAD

**Issue by zhenlan** [aspnet/MicrosoftConfigurationBuilders#119]
The AzureAppConfigurationBuilder should add a new parameter to accept the `clientId` of the user-assigned managed identity.

**Comment by yar-shukan** [comment]
Hi @zhenlan. I've originally reported the issue with App Service and User Assigned Identity 400 Bad Request.

From my understanding of the issue it comes from inner workings of `Azure.Identity` package and how it works with `System Assigned` identity versus `User Assigned` one.

Is there any chance **not to introduce this new parameter** `clientId` but just get this `clientId` from the list of App Service's User Assigned identity **if there's only one identity in the list**? 

This would keep configuration much cleaner for most of the use cases I guess.

Please let me know about your opinion. Thanks!

**Comment by zhenlan** [comment]
Hi @yar-shukan if you are requesting a design change of `ManagedIdentityCredential`, you may open an issue at https://github.com/Azure/azure-sdk-for-net. Folks from the `Azure.Identity` team will respond accordingly. The `AzureAppConfigurationBuilder` in this repo is just a consumer of `ManagedIdentityCredential`.

**Comment by StephenMolloy** [comment]
We definitely want to do the right thing for people who use this package and the App Configuration service. I worry about polluting the builder declaration with too many options resulting in an experience that is more difficult to understand and manage for the typical user. But if the typical user - or at least a not super-extreme-corner-case user - is having trouble here, we certainly want to make the experience better.

I'm curious why this isn't working though. The config builder uses [DefaultAzureCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet), which is basically a wrapper to try these four credential types in order:

- EnvironmentCredential
- ManagedIdentityCredential
- SharedTokenCacheCredential
- InteractiveBrowserCredential

Shouldn't user-assigned identities get picked up by the ManagedIdentityCredential? Perhaps I'm missing something?

**Comment by zhenlan** [comment]
You are right that the `ManagedIdentityCredential` will be picked up via `DefaultAzureCredential`, but since the `clientId` is not provided, it will only pick up the **system-assigned managed identity**. You can pass a [`DefaultAzureCredentialOptions`](https://docs.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredentialoptions?view=azure-dotnet)`.ManagedIdentityClientId` to `DefaultAzureCredential` when `clientId` is provided, and then it will be able to pick up the **user-assigned managed identity**.

**Comment by StephenMolloy** [comment]
Ah yes, I see that now. So this is something we would have to address in this package. Seems easy enough.

**Comment by yar-shukan** [comment]
@zhenlan , @StephenMolloy sorry, but I didn't get exactly what action item on this issue will be? should I create an issue in [Azure.Identity](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/identity/Azure.Identity) or how will we proceed to fix current situation that User Managed Identity doesn't work at all? thanks!

**Comment by StephenMolloy** [comment]
This is something I will try to fix in this package. The timing of this is spotty though with all the schedule and priority upheavals  in recent  weeks.

**Comment by yar-shukan** [comment]
Does that mean you will you add a new required parameter like `clientId` to make that working?

**Comment by yar-shukan** [comment]
@StephenMolloy, @zhenlan I think you can close this issue: this issue is easy fixed by using environment variable `AZURE_CLIENT_ID` set to `Client ID` of the user assigned identity.
Having this, I think we don't need the `clientId` parameter that was mentioned above.

**Comment by StephenMolloy** [comment]
Fixed with #174 and #142.

The former updates to use Azure.Identity 1.2 which adds `EnvironmentCredential` to the list of possible credential sources in the default credential we use. This can be used for assigning user-managed and client-cert based credentials. (Note, individual projects can always update the Azure.Identity package on their own to get these new capabilities without waiting for this project to update dependency requirements. And in fact this may be desirable as it is a goal of this project to _not_ force dependency updates on anyone using us without a really good reason. In this case we went up to Azure.Identity 1.2, but did not go as far as 1.4 which adds support for logging in with Azure PS through the default credential.)

For projects that need capabilities beyond what DefaultAzureCredential can provide, #142 offers the ability to provide your own credentials by sub-classing either of our Azure builders.
